### DISCUSSION TITLE
Auth headers for httpfs to securely query data via HTTP(S)

### DISCUSSION THREAD

**Discussion by tobilg** [duckdb/duckdb#5972] [Ideas]
If I understood correctly, it's currently not possible to supply auth headers for httpfs when e.g. querying parquet files via HTTP(S). This makes if close to impossible to query private data, that isn't located in S3. 

Use Case here would be for example using CloudFront to cache parquet files, and therefore using Lambda@Edge functions to validate JWTs via `Authorization` headers.

Ref. https://github.com/sutoiku/puffin/issues/6#issuecomment-1399358946

**Comment by Mause** [comment]
My hope is that this will be kinda solved for Python with https://github.com/duckdb/duckdb/pull/5829, but obviously that won't help with other languages

**Reply by tobilg** [reply]
Thanks for the fast feedback, I forgot to mention that I‘m using Node.js

**Reply by carlopi** [reply]
Could you not employ some sort of proxy to forward the requests as needed? Potentially as a stop-gap?
Surface otherwise seems pretty big to handle all kind of protocols, potentially proper way could be via pluggable extensions somehow.

**Reply by Mause** [reply]
Unfortunately we don't yet have proper proxy support: https://github.com/duckdb/duckdb/pull/4741

**Reply by tobilg** [reply]
For me, having `Authorization` or `x-api-key` headers would be enough. No need for something fancy. E.g. setting the headers could work similarly as setting the AWS credentials in the httpfs extension:

`SET http_header_authorization='Bearer foighdfioghdfiohgod...'` 
or
`SET http_header_xapikey='8dfadg66da7f6g7ad86a8d'`

**Reply by Mause** [reply]
> For me, having `Authorization` or `x-api-key` headers would be enough. No need for something fancy. E.g. setting the headers could work similarly as setting the AWS credentials in the httpfs extension:
> 
> `SET http_header_authorization='Bearer foighdfioghdfiohgod...'` 
> or
> `SET http_header_xapikey='8dfadg66da7f6g7ad86a8d'`

I think it would probably make more sense to allow setting it as a map, that way people can set whichever headers they like?

**Reply by tobilg** [reply]
That'd be even better @Mause :-) For my simple auth needs those two would be enough though...

**Reply by jakthom** [reply]
➕ to this, I would love to be able to pass headers for remote http reads.

It would also be incredibly useful when securely attaching db files -> https://duckdb.org/docs/sql/statements/attach.html

**Reply by tobilg** [reply]
It’s currently not in the docs, but have a look at https://motherduck.com/blog/duckdb-110-hidden-gems/. Haven’t tried it with ATTACH though

**Reply by sounkou-bioinfo** [reply]
> It’s currently not in the docs, but have a look at https://motherduck.com/blog/duckdb-110-hidden-gems/. Haven’t tried it with ATTACH though

@tobilg did you try with attach ? Does it work with duckdb-wasm
